torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1706,84 +1750,181 @@ struct to_ir {
+    // For use with `unifyTypeList`
+    std::stringstream nowhere;
+
+    Value* lhs_val;
+
+    // Get the types on the right-hand side of the `isinstance`
+    // expression, e.g. `isinstance(x, (float, int))` -> `{float, int}.
+    // While doing this, we also want to decompose any `Union` types,
+    // namely `Union`, `Optional`, and `Number`
+    // NB: We can't use `auto` with this lambda because it contains a
+    // recursive call
+    std::function<void(const Expr&, std::vector<TypePtr>&, bool)> gather_types;
+
+    gather_types =
+        [&](const Expr& expr, std::vector<TypePtr>& types, bool is_lhs) {
+          TypePtr t;
+
+          if (!is_lhs) {
+            if (expr.kind() == TK_TUPLE_LITERAL) {
+              for (Expr e : TupleLiteral(expr).inputs()) {
+                gather_types(e, types, is_lhs);
+              }
+              return;
+            }
+            t = typeParser_.parseTypeFromExpr(expr);
+          } else {
+            lhs_val = emitExpr(expr);
+            t = lhs_val->type();
           }
-          return;
-        }
-        TypePtr type = typeParser_.parseTypeFromExpr(classinfo);
-        types.emplace_back(type);
-      }
-      bool staticallyTrue(const TypePtr& actual_type) {
-        // is this isinstance check statically true?
-        for (const TypePtr& typ : types) {
-          if (actual_type->isSubtypeOf(typ)) {
-            return true;
+
+          if (t->kind() == UnionType::Kind) {
+            auto inner = t->expect<UnionType>()->containedTypes().vec();
+            types.insert(types.end(), inner.begin(), inner.end());
+          } else if (t->kind() == OptionalType::Kind) {
+            auto inner = t->expect<OptionalType>()->getElementType();
+            types.push_back(inner);
+            types.push_back(NoneType::get());
+          } else if (t->kind() == NumberType::Kind) {
+            types.push_back(IntType::get());
+            types.push_back(FloatType::get());
+            types.push_back(ComplexType::get());
+          } else {
+            types.push_back(t);
           }
+        };
+
+    auto dedup = [&](std::vector<TypePtr>& types) {
+      std::sort(
+          types.begin(),
+          types.end(),
+          [&](const TypePtr& t1, const TypePtr& t2) {
+            if (t1->kind() != t1->kind()) {
+              return t1->kind() < t2->kind();
+            }
+            return t1->str() < t2->str();
+          });
+
+      auto new_end = std::unique(
+          types.begin(),
+          types.end(),
+          [&](const TypePtr& t1, const TypePtr& t2) { return *t1 == *t2; });
+
+      types.erase(new_end, types.end());
+    };
+
+    std::vector<TypePtr> lhs_types;
+    std::vector<TypePtr> rhs_types;
+
+    gather_types(obj, lhs_types, /*is_lhs=*/true);
+    gather_types(classinfo, rhs_types, /*is_lhs=*/false);
+
+    RefinementSet refinement;
+
+    TypePtr unified_true = nullptr;
+    TypePtr unified_false = nullptr;
+
+    if (obj.kind() == TK_VAR) {
```

Comment:
we only do this logic for `TK_VAR`, but don't we also want to do the filtering of isinstance types even for unnamed expressions ?

Reply:
I might be misreading the original code, but it seems like that's what we do [there](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/ir_emitter.cpp#L1742), as well. The `RefinementSet` `refinement` will only have something in it if the lhs is a `Var`